src/stream/src/executor/source/source_executor.rs

```diff
@@ -245,27 +245,57 @@ impl<S: StateStore> SourceExecutor<S> {
             .stack_trace("source_recv_first_barrier")
             .await
             .unwrap();
+        let epoch = barrier.epoch.prev;
 
         if let Some(mutation) = barrier.mutation.as_ref() {
             if let Mutation::Add { splits, .. } = mutation.as_ref() {
                 if let Some(splits) = splits.get(&self.ctx.id) {
                     self.stream_source_splits = splits.clone();
+                    for split_impl in &mut self.stream_source_splits {
+                        if let Some(recover_state) = self
+                            .split_state_store
+                            .try_recover_from_state_store(split_impl.id(), epoch)
+                            .await?
+                        {
+                            *split_impl = recover_state;
+                        }
+                    }
+
+                    // save source_id -> [split_id] to state store
+                    self.split_state_store
+                        .save_split_assignment(splits, epoch)
+                        .await?;
                 }
             }
+        } else {
```

Comment:
Is it also possible that the mutation of the first barrier is `Mutation::Update`, which occurs after scaling out and also contains some split assignments? Similar for `Mutation::SplitChange`? cc @shanicky

Reply:
The first barrier can only be `Mutation::Add` (for create) or `Mutation::Update`(for scale), others should be unreachable.